### PR TITLE
Tests for CreateTestRun views refractored

### PR DESCRIPTION
**Pull request by SvetlomirBalevski**
Add tests for CreateTestRun view using new PermissionsTestCase class.

### CODE REVIEW

File: tcms/testruns/tests/test_permissions.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    def verify_get_with_permission(self):
+        response = self.client.get(self.url)
+
+        self.assertEqual(HTTPStatus.OK, response.status_code)
+
+    def verify_post_with_permission(self):
+
+        response = self.client.post(self.url, self.post_data, follow=True)
+
+        self.assertEqual(HTTPStatus.OK, response.status_code)
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: See what `bugs/tests/test_permissions.py::NewTestCase.verify_post_with_permission` does as well as 
`testruns/tests/test_views.py::TestCreateNewRun.test_create_a_new_run()`. You have partially copied some of that code and left the rest.

You will have to copy all of the asserts and adjust the test case accordingly.

Notice how the original test case doesn't use `follow=True`

File: tcms/testruns/tests/test_permissions.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+from tcms import tests
+from tcms.tests import factories
+from tcms.testcases.models import TestCaseStatus
+
+
+class TestCreateTestRun(tests.PermissionsTestCase):
+    permission_label = 'testruns.add_testrun'
+    url = reverse('testruns-new')
+    http_method_names = ['get', 'post']
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: The view under test doesn't support GET requests.

Note: I get a 200 in browser instead of 4xx, not sure why ATM.

File: tcms/testruns/tests/test_permissions.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            author=cls.tester,
+            default_tester=None,
+            reviewer=cls.tester,
+            case_status=cls.case_status_confirmed,
+            plan=[plan])
+        case_2.save()  # will generate history object
+
+        cls.post_data['case'] = (case_1.pk, case_2.pk)
+
+    def verify_get_with_permission(self):
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: that isn't necessary wrt get vs post statement above.

File: tcms/testruns/tests/test_permissions.py (hunk 1 above)
Comment (atodorov) [reply tcms/testruns/tests/test_permissions.py]: Note: this view does support 2 modes of operation based on POSTING_TO_CREATE variable. I need to examine this in more details to see what is going on there.

**Comment by atodorov** [comment]
Also fix pylint errors in your code.

File: tcms/testruns/tests/test_permissions.py (hunk 2 above)
Comment (atodorov) [reply tcms/testruns/tests/test_permissions.py]: This is still an issue but you failed to see b/c of the missing changes introduced in #1196. Once that is merged rebase and this will start erroring out.

File: tcms/testruns/tests/test_permissions.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            default_tester=None,
+            reviewer=cls.tester,
+            case_status=case_status_confirmed,
+            plan=[plan])
+        case_2.save()  # will generate history object
+
+        cls.post_data['case'] = (case_1.pk, case_2.pk)
+
+    def verify_post_with_permission(self):
+        new_run = TestRun.objects.last()
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Two things here:

1) When moving code around delete the origin in `testruns/tests/test_views.py::TestCreateNewRun.test_create_a_new_run()`, otherwise you end up with duplciates

2) The new test case is not a verbatim copy of the old. A simple copy &  paste here doesn't work, you have to think about what are the important things we want to be checking(aka verifying) in this scenario. Again analyze `bugs/tests/test_permissions.py::NewTestCase.verify_post_with_permission` for hints.

**Comment by atodorov** [comment]
@RMadjev beat you to adding tests in this file so rebase when you get back to this.

**Comment by atodorov** [comment]
ping, any update here ?

**Comment by SvetlomirBalevski** [comment]
> ping, any update here ?

Need to fix the redirect test.

AssertionError: '/runs/1/' != '/accounts/login/?next=%2Fruns%2F1%2Fedit%2F'
- /runs/1/
+ /accounts/login/?next=%2Fruns%2F1%2Fedit%2F
 : Response redirected to '/runs/1/', expected '/accounts/login/?next=/runs/1/edit/'Expected '/runs/1/' to equal '/accounts/login/?next=/runs/1/edit/'.

File: tcms/testruns/tests/test_permissions.py [hunk 5]
```diff
@@ -56,3 +58,39 @@ def verify_post_with_permission(self):
         self.assertRedirects(
             response,
             reverse('testruns-get', args=[self.test_run.pk]))
+
+
+class TestRunTestCase(tests.PermissionsTestCase):
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: -> `CreateTestRunViewTestCase` (convention is: name of view under test + "TestCase")

File: tcms/testruns/tests/test_permissions.py [hunk 6]
```diff
@@ -56,3 +58,39 @@ def verify_post_with_permission(self):
...
+    permission_label = 'testruns.add_testrun'
+    url = reverse('testruns-new')
+    http_method_names = ['post']
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.test_run = factories.TestRunFactory(summary='Old summary')
+        cls.url = reverse('testruns-edit', args=[cls.test_run.pk])
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Why override the `url` attribute? 

`cls.url` here and the `url` on line 65 is exactly the same thing.

File: tcms/testruns/tests/test_permissions.py [hunk 7]
```diff
@@ -2,11 +2,13 @@
 # pylint: disable=invalid-name, too-many-ancestors
 
 
-from django.urls import reverse
 from django.utils.translation import ugettext_lazy as _
 
 from tcms.tests import PermissionsTestCase
+from django.urls import reverse
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Why did you rearrange the imports ? Was there a specific reason for that ? 

If not check your pylint output!

File: tcms/testruns/tests/test_permissions.py [hunk 8]
```diff
@@ -56,3 +58,39 @@ def verify_post_with_permission(self):
...
+
+        intern = factories.UserFactory(username='intern',
+                                       email='intern@example.com')
+        cls.post_data = {
+            'summary': 'New run summary',
+            'build': cls.new_build.pk,
+            'manager': cls.test_run.manager.email,
+            'default_tester': intern.email,
+            'notes': 'New run notes',
+        }
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: This `post_data` doesn't seem correct. Why did you delete the previous code and substitute with something completely unrelated ?  See previous comments in this section.

File: tcms/testruns/tests/test_permissions.py [hunk 9]
```diff
@@ -56,3 +60,101 @@ def verify_post_with_permission(self):
...
+        cls.post_data = {
+            'summary': plan.name,
+            'from_plan': plan.pk,
+            'build': build.pk,
+            'notes': 'Create new test run',
+            'POSTING_TO_CREATE': 'YES',
+        }
+        cls.tester = factories.UserFactory()
+        cls.tester.set_password('password')
+        cls.tester.save()
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: this is already defined in the parent class, why do you need it ?

File: tcms/testruns/tests/test_permissions.py [hunk 10]
```diff
@@ -56,3 +60,101 @@ def verify_post_with_permission(self):
...
+        plan = factories.TestPlanFactory()
+        build = factories.BuildFactory()
+        cls.product = ProductFactory(name='Kiwi')
+
+        cls.build_fast = factories.BuildFactory(name='fast', product=cls.product)
+
+        cls.post_data = {
+            'summary': plan.name,
+            'from_plan': plan.pk,
+            'build': build.pk,
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Why build instead of build_fast ?

File: tcms/testruns/tests/test_permissions.py [hunk 11]
```diff
@@ -56,3 +60,101 @@ def verify_post_with_permission(self):
...
+        cls.case_2 = factories.TestCaseFactory(
+            author=cls.tester,
+            default_tester=None,
+            reviewer=cls.tester,
+            case_status=case_status_confirmed,
+            plan=[plan])
+        cls.case_2.save()  # will generate history object
+
+        cls.post_data['case'] = (cls.case_1.pk, cls.case_2.pk)
+        super(CreateTestRunViewTestCase, cls).setUpTestData()
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Use python3 super()

File: tcms/testruns/tests/test_permissions.py [hunk 12]
```diff
@@ -56,3 +60,101 @@ def verify_post_with_permission(self):
...
+        clone_data = {
+            'summary': self.plan.name,
+            'from_plan': self.plan.pk,
+            'build': self.build_fast.pk,
+            'manager': self.tester.email,
+            'default_tester': self.tester.email,
+            'notes': 'Clone new run',
+            'case': [self.case_1.pk, self.case_2.pk],
+            'POSTING_TO_CREATE': 'YES',
+        }
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Why define test data again ?

File: tcms/testruns/tests/test_permissions.py [hunk 13]
```diff
@@ -56,3 +58,79 @@ def verify_post_with_permission(self):
...
             reverse('testruns-get', args=[self.test_run.pk]))
+
+
+class CreateTestRunViewTestCase(tests.PermissionsTestCase):
+    permission_label = 'testruns.add_testrun'
+    http_method_names = ['post']
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.url = reverse('testruns-new')
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: This can be defined above together with permission_label.

File: tcms/testruns/tests/test_permissions.py [hunk 14]
```diff
@@ -56,3 +58,79 @@ def verify_post_with_permission(self):
...
+            default_tester=None,
+            reviewer=cls.tester,
+            case_status=case_status_confirmed,
+            plan=[cls.plan])
+        cls.case_2.save()  # will generate history object
+
+        cls.post_data['case'] = [cls.case_1.pk, cls.case_2.pk]
+
+    def verify_post_with_permission(self):
+        url = reverse('testruns-new')
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: do not use stand alone variable, use the existing class attribute instead

File: tcms/testruns/tests/test_permissions.py [hunk 15]
```diff
@@ -56,3 +58,79 @@ def verify_post_with_permission(self):
...
+
+        response = self.client.post(url, self.post_data)
+        print(response.content)
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Remove print

File: tcms/testruns/tests/test_permissions.py [hunk 16]
```diff
@@ -56,3 +58,79 @@ def verify_post_with_permission(self):
...
+        new_run = TestRun.objects.last()
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: rename to `last_test_run` or `last_run`, etc

File: tcms/testruns/tests/test_permissions.py [hunk 17]
```diff
@@ -56,3 +58,79 @@ def verify_post_with_permission(self):
...
+        self.assertEqual(self.build_fast, new_run.build)
+        self.assertEqual(self.tester, new_run.manager)
+        self.assertEqual(self.tester, new_run.default_tester)
+        for case, case_run in zip((self.case_1, self.case_2),
+                                  new_run.case_run.order_by('case')):
+            self.assertEqual(case, case_run.case)
+            self.assertEqual(None, case_run.tested_by)
+            self.assertEqual(self.tester, case_run.assignee)
+            self.assertEqual(TestExecutionStatus.objects.get(name='IDLE'),
+                             case_run.status)
```
Comment (atodorov) [inline tcms/testruns/tests/test_permissions.py]: Remove this assert b/c it make things uglier wrt upcoming changes in this area
